### ISSUE TITLE
iOS - Chromecast

### ISSUE THREAD

**Issue by LePips** [jellyfin/Swiftfin#271]
**Describe the feature you'd like**
Chromecast that was previously implemented was not to its full potential. Re-implement such that the casting framework is passively running in the background and that a stream can be made without starting playback. Normal streaming during playback as well.

Also implement listening to remote events from the cast: https://github.com/jellyfin/Swiftfin/issues/177

**Comment by Jellman01** [comment]
Would love a chrome cast implementation. Signed up to the TestFlight, will test once in.

**Comment by kylecooper-me** [comment]
To add to this - the Cast icon should ALWAYS be present.

For some reason in the new update it’s just totally disappeared on me and I can’t chromecast.

Please literally copy Plex when it comes to chromecast support - I shouldn’t have to press “play” and then swear under my breath when the cast icon decides to not appear.

**Comment by anthonylavado** [comment]
> To add to this - the Cast icon should ALWAYS be present.
>...
>Please literally copy Plex when it comes to chromecast support - I shouldn’t have to press “play” and then swear under my breath when the cast icon decides to not appear. Just let me press Cast from the Home Screen.

That's the goal of this re-implementation.

> For some reason in the new update it’s just totally disappeared on me and I can’t chromecast.

It's been taken out because it has to be re-done, to enable the things you mentioned. Apologies for that.

**Comment by kylecooper-me** [comment]
> That's the goal of this re-implementation.

Thats great, and I'm all for progress.

> It's been taken out because it has to be re-done, to enable the things you mentioned. Apologies for that.

The beta app with Chromecast support is the only way I can use Jellyfin. Removing Chromecast support outright means I need to go back to using Plex.

I'd like to see the working 'stableish' Swiftfin version remain available so that people like myself have a way of watching TV shows without needing to go back to Plex.

**Comment by anthonylavado** [comment]
>I'd like to see the working 'stableish' Swiftfin version remain available…

I'll ask the current publisher if we can still make that build available. Within TestFlight, it's possible to keep previous builds or expire them when upgrading. I'm not sure if the previous one can still be restored, but we'll see.

>Now I'd argue that Swiftfin is objectively better than the Jellyfin app itself and is a worthy replacement - despite its shortcomings. Perhaps a solution would be to replace Jellyfin app with Swiftfin-stable with Chromecast support.

We'll likely have both around for some time. The "main" app has the full feature set at the moment, including Live TV, Music, and the Admin dashboard. While Live TV and Music should eventually carry over to this app, the Admin dashboard may be tricky for some time, but that's another story. We wouldn't be replacing one with the other just yet.

**Comment by 0x6b-dev** [comment]
Bit the bullet and just bought a Google TV (begrudgingly). Though very happy with the experience.

IMHO - theres something to be said about springing feature regressions onto users without a tangible alternative. If Jellyfin's solution was "suck it up and go buy a Google TV" id have accepted that - but its pretty frustrating sitting down to watch TV one night and realising you need to spin up your old Plex container to do so.

I digress - is there a way that a non-developer can contribute (financially or otherwise) now that I'm free of the clutches of proprietary media services?

**Comment by ncallaway** [comment]
@LePips Is this something that a new contributor would be able to pick up, or do you have specific plans for how this should be implemented and want to tackle it yourself?

Or is this a task that can be decomposed into smaller tasks, and have contributors help with individual pieces?

**Comment by LePips** [comment]
Sorry it's been a while and thanks for your interest in getting this along! I've done work for this before however it's been so long that I've had a bit to think about its implementation, especially since I've also worked with downloading.

I have a lot in mind. We are going to have _at least_ three different forms of video playback: normal, offline, and Chromecast. I need to better formalize how streams are handled which will require a lot of reworking and new writing. I don't have a concrete plan for what distinct things need to happen so I sadly won't be able to break it up.

I'm currently working through https://github.com/jellyfin/Swiftfin/pull/452, but this is the next thing I plan on tackling as well as downloading. Then, I personally feel like the app is at a good point to be officially launched.

**Comment by ncallaway** [comment]
Makes sense! I could see it being a hard thing to hand off to a new contributor.

As you get into it, let me know if there's anything you find you want to hand off. Obviously, no worries if it's a one person job when you get into it

**Comment by Norrox** [comment]
Sorry to bump this, I saw the app was live in the AppStore, but I do not have the option to cast to my chromecast here at home. Is it not implemented? 🙂

**Comment by emmc15** [comment]
Following up on the Bump above. Wondering if there is any plan to further look into this or open it up to pass it along to someone else if they were interested.

Love the app! Thanks for all the contributions, having Chromecast feature would just make such a killer feature!

**Comment by fivestones** [comment]
I'm wondering the same as @emmc15 above. Is anything still in the works for this?

**Comment by citric-code** [comment]
I would also like to bump this, to see if anyone has any solution to adding Chromecast, was very infuriating to find my wife cannot cast to any TV in our house after spending ages setting up a server.

**Comment by CandiMiiS** [comment]
Back to plex it is.. no chromecast support is a major fault

**Comment by weiyideai520** [comment]
There is still a way to join the TestFlight now, I need Chromecast.

**Comment by ReclinesMC** [comment]
@LePips  did this ever become a thing? :P

**Comment by thornbill** [comment]
If there were any updates you would see that here. No need to keep asking.

**Comment by Blakem96** [comment]
Hi folks—I'd like to take a first pass at adding Google Cast support to Swiftfin.
Plan: implement a minimal, feature-flagged MVP (default off) so it’s safe to merge incrementally.
MVP includes: device discovery, connect/disconnect, play/pause/seek and now-playing metadata.
Non-goals for MVP: track/subtitle selection, queue management, Live TV, trickplay, multiple users, or deep settings.
I’ll open a draft PR early for feedback.
If maintainers are okay with this direction, please assign me or just consider this “in progress.”
Figure working on baby steps is better than no progress at all :)

**Comment by LePips** [comment]
I really don't mean to discourage such work, but I can't recommend doing so without understanding the new manager architecture in #1581. It would also be a lot more work than making the player with other feature requirements.

I worked a little bit on it during #1581 and may get to it before you do.
